File: proto/src/cli/ui.ts
import * as p from "@clack/prompts";
import { render } from "ink";
import React from "react";
import { Box, Text } from "ink";
import { App } from "../tui/App";
import { readConfig, type WorkspaceConfig } from "../config";
import { Store, unlock } from "../store";
import { generateInvite } from "../invite";
import { derivePrivateKey } from "../crypto";

export async function cmdUi() {
  const config = await readConfig();

  if (!config || config.workspaces.length === 0) {
    console.error("No workspaces found. Run: bkey setup");
    process.exit(1);
  }

  let workspace: WorkspaceConfig;

  if (config.workspaces.length === 1) {
    workspace = config.workspaces[0] as WorkspaceConfig;
  } else {
    const selected = await p.select({
      message: "Select workspace",
      options: config.workspaces.map((w) => ({ value: w.id, label: w.name })),
    });

    if (p.isCancel(selected)) {
      p.cancel("Cancelled.");
      process.exit(0);
    }

    workspace = config.workspaces.find(
      (w) => w.id === selected,
    ) as WorkspaceConfig;
  }

  const store = new Store({
    memberId: config.memberId,
    workspaceId: workspace.id,
    storage: workspace.storage,
  });

  // Pull and unlock before ink takes over the terminal
  const doc = await store.pull();
  const workspaceId = doc.id;

  const privateKey = derivePrivateKey(config.passphrase, workspaceId);

  const { rerender, unmount } = render(
    React.createElement(
      Box,
      { paddingX: 1 },
      React.createElement(Text, { dimColor: true }, "Loading secrets…"),
    ),
  );

  try {
    const { session, doc: updatedDoc } = await unlock(doc, privateKey);
    const inviteLink = await generateInvite(workspace.storage, {
      id: doc.id,
      name: doc.name,
    });

    rerender(
      React.createElement(App, {
        initialDoc: updatedDoc,
        store,
        session,
        inviteLink,
      }),
    );
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    rerender(
      React.createElement(
        Box,
        { paddingX: 1, gap: 1 },
        React.createElement(Text, { color: "red" }, "✗"),
        React.createElement(Text, null, message),
      ),
    );
    unmount();
    process.exit(1);
  }
}

File: envi/cli/src/commands/ui.rs
use envilib::{
    config::read_config,
    crypto::derive_private_key,
    error::{Error, Result},
    invite::{generate_invite, WorkspacePayload},
    store::{unlock, Store},
};

pub async fn run() -> Result<()> {
    let config = read_config().await?.ok_or(Error::NoConfig)?;

    if config.workspaces.is_empty() {
        return Err(Error::NoWorkspaces);
    }

    let workspace = if config.workspaces.len() == 1 {
        config.workspaces.into_iter().next().unwrap()
    } else {
        let names: Vec<_> = config.workspaces.iter().map(|w| w.name.as_str()).collect();
        let idx = dialoguer::Select::new()
            .with_prompt("Select workspace")
            .items(&names)
            .default(0)
            .interact()
            .map_err(|e| Error::Other(e.to_string()))?;
        config.workspaces.into_iter().nth(idx).unwrap()
    };

    let store = Store::new(&workspace.id, &config.member_id, &workspace.storage)?;
    let doc = store.pull().await?;
    let private_key = derive_private_key(&config.passphrase, &workspace.id)?;
    let session = unlock(&doc, &private_key)?;

    let invite_link = generate_invite(
        &workspace.storage,
        WorkspacePayload {
            id: workspace.id.clone(),
            name: workspace.name.clone(),
        },
    )?;

    crate::tui::run(doc, store, session, invite_link).await
}
